torch/_inductor/compile_fx_async.py

```diff
@@ -179,3 +179,183 @@ def callback(pickled_output: _WireProtocolPickledOutput) -> OutputCode:
             return output.graph
 
         return _AsyncOutputCode(eager_output_code, f, callback)
+
+
+# _ProgressiveOutputCode handles running a fast compile first, then hot-swapping
+# to a more optimized version when the expensive compile finishes.
+@final
+class _ProgressiveOutputCode(OutputCode):
+    _fast_output_code: Optional[OutputCode]
+    _optimized_output_code: Optional[OutputCode]
+    _progression_futures: list[Future[_WireProtocolPickledOutput]]
+    _callback: Callable[[_WireProtocolPickledOutput], OutputCode]
+    _post_compile_data: Optional[_PostCompileData] = None
+    _boxed_call: bool
+    _current_progression_index: int
+
+    def __init__(
+        self,
+        # Fast compile that runs faster than the progressive compiles
+        fast_output_code: OutputCode,
+        # Futures for the progressive optimized compiles
+        progression_futures: list[Future[_WireProtocolPickledOutput]],
+        # Callback to convert the optimized result to OutputCode
+        callback: Callable[[_WireProtocolPickledOutput], OutputCode],
+    ) -> None:
+        self._fast_output_code = fast_output_code
+        self._optimized_output_code = None
+        self._progression_futures = progression_futures
+        self._callback = callback
+        self._boxed_call = getattr(fast_output_code, "_boxed_call", False)
+        self._current_progression_index = -1
+
+    @override
+    def __call__(self, *args: Any) -> Any:
+        # Check if any newer progression stage is ready and switch to it
+        args = self._check_and_switch_progression(args)
+
+        if self._optimized_output_code is not None:
+            _ProgressiveFxCompile._stat_optimized_runs += 1
+            return self._optimized_output_code.__call__(*args)
+        else:
+            _ProgressiveFxCompile._stat_fast_runs += 1
+            assert self._fast_output_code is not None
+            return self._fast_output_code.__call__(*args)
+
+    def _check_and_switch_progression(self, args: tuple[Any, ...]) -> tuple[Any, ...]:
+        # Check if any newer progression stage is ready (in order from latest to earliest)
+        for i in range(
+            len(self._progression_futures) - 1, self._current_progression_index, -1
+        ):
+            future = self._progression_futures[i]
```

Comment:
nit:
```suggestion
        for i, future in reversed(enumerate(self._progression_futures)):
```

Reply:
This is technically less efficient right?

- enumerate() returns an iterator
- reversed() requires a sequence (not just an iterator), so it converts the iterator to a list internally
- Then it reverses that materialized list

whereas range uses an iterator with no materialization? I guess the big O is negligible and we get readability benefits?

Reply:
Yeah - I was going for readability. But apparently you can't do `reversed(enumerate(...))` (you need to convert to a list) so that makes it less appealing...